Update CI

Cimbali opened the pull request:
I know @shymega had a look at this branch and preferred sticking with `clean_build.bat` approaches rather than what I did here, so I tried to convert this working CI back to the script-based form now I had ironed out the kinks.

However, I quickly ran into all the reasons I inlined the scripts in the first place, listed below (a bit verbose maybe, sorry), so I still wanted to open the PR in this state so we can discuss what is and what isn’t desirable.

Overall I think this *seems* to add more complexity, but really makes the CI more maintainable, and especially easier to read for non-insiders, meaning it will be easier to get the community to help you on it.

-------

- More oversight into what happens  
  - all commands/flags/values are echoed to the CI
    
    ![image](https://user-images.githubusercontent.com/6126377/229287143-016d6460-27b9-46a9-bac1-9684fe6663b5.png)

  - In case of an error, instead of having an error happen in a script, now a specific command throws an error
    
    ![image](https://user-images.githubusercontent.com/6126377/229287214-f199430e-842a-4bc2-9d75-d6538ce33118.png)
     (This error is fixed in the PR)

  - A single ”source of Truth” for environment variables.  
    For example, when building through `clean_build.bat` the Qt version is set to `5.11.1` in that script, `5.15.2` in the `build_env.bat` template, and `5.15.2` in `download_install_qt.ps1` -- but those values must be kept in sync manually and it is not immediately clear which is the one that is used.

    In the PR’d CI, values are set a single time:
    https://github.com/input-leap/input-leap/blob/5972a4a2ca7f1db75032f8a4ff8436f5b1a0df8d/.github/workflows/builds.yml#L237-L239
    Then exported and reused later from that single definition:
    https://github.com/input-leap/input-leap/blob/5972a4a2ca7f1db75032f8a4ff8436f5b1a0df8d/.github/workflows/builds.yml#L267-L274

- In terms of complexity, I don’t think it’s very high, and it is kept simpler by using the same structure for all platforms:
  - `$GITHUB_WORKSPACE/input-leap` for code
  - `$GITHUB_WORKSPACE/build` for building
  - A single `cmake` configuring command per platform
     - scripts getting environment information are kept (`input-leap/macos_environment.sh`, `VsDevCmd.bat`), always in the configuring `cmake` step
     - this also makes it easier to adjust flags for the CI, see 5972a4a2ca7f1db75032f8a4ff8436f5b1a0df8d
  - A single `cmake --build` building command per platform


- Additional sugar:
  - Can be manually run on-demand
  - Single checkbox to make Release builds
  - Update versions of used actions (to avoid deprecation notices)
  - Use actions built-in options to reduce boilerplate code (recursive submodules, safe directory, etc.)
  - Matrix approach to build wayland dependencies makes conditional steps very readable, e.g. `if matrix.wayland`
  - Checkout of specific version can be set with `ref` 

- Main drawbacks:
  - Overall longer than previous version
     - Most of the length are the various dependency installing/fetching (installing dependencies, plus libei, libportal, qt, bonjour, …)
  - Syntax to conditionally set environment variables is `condition && <value if true> || <value if false>` which is not the most readable and `<value is true>` must never be falsy -- but this is commented as appropriate

## Contributor Checklist:

* [x] No user-visible changes

shymega commented:
Hey @Cimbali! REALLY good PR here. Love the design, and I have no objections.

The only thing I'd like to see added is a CI workflow using `clang`, to address a concern. Would this be possible? If that's added, I'd be willing to merge, but I do need to await for Povilas (@p12tic)'s sign-off to be honest. I have emailed him to check in, but haven't heard back. If I don't hear back in 1.5 weeks, I'll go ahead and merge.

Thank you!

shymega commented:
@Cimbali Another thing - you might have seen the post in Discussions about compiling Mac universal binaries. It worked on ARM64 and x86_64 (but might need to be separate). Do you think you could incorporate this into the CI in a subsequent PR once this one is merged?

I still haven't heard from @p12tic, I'll ping him again, but as soon as you add clang support, or it could be separate, I'm happy to merge.

Cimbali commented:
Added the clang pipeline to the matrix, and fixed a couple of things along the way, but there are still warnings that cause clang to fail (with warnings and `-Werror`).

I’m not sure I’ve seen the post about universal binaries, but it should certainly be doable too.

p12tic commented:
@Cimbali Could the warnings fixes be split off to separate PR so that they can be merged sooner? Thanks a lot!

Cimbali commented:
Sure, but you’ll get a failing CI in this PR.

Cimbali commented:
> you might have seen the post in Discussions about compiling Mac universal binaries. It worked on ARM64 and x86_64 (but might need to be separate).

I think if we want to build ARM/M1 macOS builds (which is the conversation I’ve seen), the easiest is to have a runner with the correct architecture. Unfortunately it seems those are not available from Github as of yet.

Otherwise we’d need to cross-compile but I have no idea how doable that is. But from the [qBittorrent macOS building docs](https://github.com/qbittorrent/qBittorrent/wiki/Compilation:-macOS-(x86_64,-arm64,-cross-compilation)) it seems doable (qBittorrent also depends on Qt 5.15 and OpenSSL), although requiring a lot of compiling from source.

From [the universal binary docs](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary/) it seems a universal binary are just different native binaries bundled together. That can be done in the CI even if the original builds are on different runners, with a separate job that depends on macOS building jobs, downloads the assets, and bundles them together.

p12tic commented on .github/workflows/builds.yml:
```diff
@@ -1,40 +1,47 @@
-  B_BUILD_DIR: build
-  B_BUILD_TYPE: Debug
-  INPUTLEAP_VERSION_STAGE: Debug
+  B_BUILD_TYPE: ${{ (inputs.build_release || startsWith(github.ref, 'refs/tags/v')) && 'Release' || 'Debug' }}
```
Is it possible to check if `build_release` is set at all? If yes, I think it makes sense to override `refs/tags/v` which whatever value that is set by `build_release` variable.

Cimbali replied:
Currently `build_release` is a boolean.

The logic here is build `Release` if the build is triggered by a release (triggered event is a tag, and tag starts with `v`) or `Release` checkbox is ticked.

In any case it’s the version defined in the code that is used for the build.

Cimbali replied:
The logic can be adjusted of course. Just the manual setting? Just on tags pushed? Something else?

shymega replied:
I like the existing solution, personally, and don't want to override any `refs/tags/v`. @Cimbali's solution makes sense from a CI/CD perspective, and is logical.

p12tic replied:
Sorry for the delay. @shymega this basically means that it's not possible to create a debug build on v1.2.3 tag for example. Do we really want that? I think it makes sense to leave it as an option.

Cimbali replied:
I’ll double check, but I believe you can still do that, by triggering a build on a tag manually (so the condition “triggered event is a tag” is false) and leaving the “release” checkbox unticked.

shymega replied:
@p12tic Right. Well, let's wait for @Cimbali to check, and then discuss further after they come back to us.

shymega replied:
@Cimbali Did you get a chance to double check?

cc: @p12tic

shymega commented on .github/workflows/builds.yml:
```diff
@@ -136,88 +177,151 @@ jobs:
-          python-version: '3.7'
-          architecture: 'x64'
-      - name: Download Bonjour SDK-like
-        run: ./azure-pipelines/download_install_bonjour_sdk_like.ps1
```
Yeah, this was something I raised before, Povilas. I don't quite understand your proposal, though.

Cimbali replied:
To me this moves the configuration from one CI (azure with scripts) to another CI (github yml). The [windows build instructions](https://github.com/input-leap/input-leap/wiki/Building-on-Windows) direct users to download Qt/Bonjour/etc from the respective sources, so I guess this comment only applies to the `clean_build.bat` not `download_install_bonjour_sdk_like` and `download_install_qt`?

shymega replied:
Yeah, and we could just do away with Azure, like, completely. I have no objection to duplicating build configs *now* that I've seen the implementation. It's clean, seems easy to maintain, and I don't think it's that big an issue. The azure-pipelines scripts aren't used by users as @Cimbali states, so that's not an issue.. right?

shymega replied:
I'm in favour of duplicate build configs. Let's not make the scripts anymore difficult to manage, and modifying the scripts will add complexity and increase maintenance.

shymega replied:
@p12tic Any further thoughts?

p12tic replied:
`clean_build.sh` are not for users, but an easy way to build the project for developers. The fact that we're duplicating configs means that clean_build.sh is not right for the purpose, but we can change the script as well.

shymega replied:
@p12tic Sure, but users do use it too. I don't think duplicating configs is too big of an issue. I'd really like to get this PR merged, to be honest, and how often do we change the scripts or CI? AFAIK, not particularly often, which means maintainability isn't a massive concern.. right?

shymega replied:
OK, so before I push my rebase+squash, I just wanted to resolve this conversation, @p12tic.

I saw you approved the PR, but this review is outstanding. Another thing we should arrange is the cessation of Azure Pipelines - most people I see downloading nightlies _are_ using GH Actions.

We don't need two CIs. @Cimbali's work on this PR has been very well done, and we can cope with just one CI service. It will simplify things, and I have great confidence in the stability of this workflow.

Can we arrange for this PR to be merged? We have people asking about Windows installers, and macOS DMGs - and this will resolve that.

shymega replied:
@p12tic We need to make a decision. More cases of confusion r.e Azure and GH Actions. We need this review resolved so that GitHub will not complain when merging.

shymega replied:
I'm gonna go ahead and resolve this conversation. @p12tic, if you want to revisit the config in future, and duplication, we can definitely do that.

But as we've all approved the PR, I'm going to a) resolve this convo and b) merge the PR into master.

I hope that's OK.

shymega commented:
@p12tic I'd like to get this merged soon. Just wanted to give you a prod.

shymega commented:
We could also look into disabling Azure Pipelines once this PR is merged.

shymega commented:
I can't use Azure Pipelines because of my Microsoft 365 Business setup...

shymega commented:
I've rebased this PR against the latest changes in mainline, and bumped `libei` after #1643 - plus added the new Python dependencies. Would be nice to get this merged soon...

shymega reviewed (approved):
I really like this approach to the current revision. It looks logical, well-designed, and I have no objection to duplicating `clean_build.sh`. It also has good commenting. I approve of this PR and have no qualms about merging it into mainline.

shymega commented:
Rebased again. libei has been bumped to 0.99.1.

shymega commented:
Had to amend the last commit, and fix a missing dependency. All done now. @p12tic @Cimbali should we get a discussion going on the merging of this PR?

whot commented on .github/workflows/builds.yml:
```diff
@@ -73,59 +81,93 @@ jobs:
-            git clone --depth 1 --branch 0.99.1 https://gitlab.freedesktop.org/libinput/libei
-            cd libei
-            meson -Dprefix=/usr -Dtests=disabled -Dliboeffis=disabled -Ddocumentation=[] _libei_builddir
+            meson -Dprefix=/usr -Dtests=disabled -Dliboeffis=disabled -Ddocumentation=[] libei _libei_builddir
```
```suggestion
            meson setup -Dprefix=/usr -Dtests=disabled -Dliboeffis=disabled -Ddocumentation=[] libei _libei_builddir
```
using just `meson` is now deprecated, it's always `meson setup` now

Cimbali replied:
Meson 1.0.1 indeed prints
```
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
```

whot reviewed (commented):
ftr, this LGTM, at least the linux bits. No idea on the windows/macos bits though...

Cimbali commented:
> Had to amend the last commit, and fix a missing dependency. All done now. @p12tic @Cimbali should we get a discussion going on the merging of this PR?

I don’t think I have much new to add to the discussion at this point. I think all the arguments have been laid out and we may be lacking a consensus? Not sure what the usual way of reaching a decision for these sort of situations is, in this project.

shymega commented:
Yeah. I think, the only change I'd like to see before merging (don't do it yet, until we get approval from Povilas), is squashing all commits to one singular commit.

@p12tic We really could do with merging this PR ASAP - I'm getting requests on how to install IL, and the installer doesn't seem to be working on Windows. I know you're busy, but I don't see issues with this PR and would like to merge it.

Can we have a look this weekend?

shymega commented:
We should also disable Azure Pipelines. I can't manage the Azure pipeline with my M365 Business account, and this PR sorts out GH Actions to the point, that we _can_ use it alone.

shymega commented:
@Cimbali You may have seen I mentioned you in the Qt 6 PR. Do you think we could remove Ubuntu 18.04 from CI if @p12tic wants to? It is EOL after all.

Cimbali commented:
Removing ubuntu 18 from the CI would be very easy to do.

shymega commented:
@p12tic @Cimbali I've just implemented the `libei` depbump from @whot. Are we on track to merging this PR?

Cimbali commented:
It’s all good to go imho.

shymega commented:
Yeah, we could remove Ubuntu 18.04 before merging, but that might be able to wait until before the Qt 6 PR is merged.

shymega commented:
Two more things as an afterthought: I was thinking of using the release functionality of this workflow to automagically trigger new releases on the Snap, Chocolately, Scoop, and Flatpak - these are registered, but not uploaded yet. Could we achieve this with a workflow trigger, or should I setup a AWS Lambda?

Cimbali commented on .github/workflows/builds.yml:
```diff
@@ -136,88 +178,151 @@ jobs:
-        if: ${{ env.CI_ENV_BUILD_TYPE == 'Release' }}
+        if: matrix.os == 'windows-2022' && env.B_BUILD_TYPE == 'Release'
         with:
           name: Publish the Installer
```
That’s correct. Now the zip name does not contain `matrix.os`, so if there were several matrix configurations for which we run this upload, all the installers would end up in the same zip, though the doc [do not recommend it](https://github.com/actions/upload-artifact#uploading-to-the-same-artifact).

Cimbali replied:
I’ve simplified the zip name.

Cimbali commented on .github/workflows/builds.yml:
```diff
@@ -136,88 +178,151 @@ jobs:
       - uses: actions/upload-artifact@v3
+        if: env.B_BUILD_TYPE == 'Release'
         with:
           name: Publish the DMG for ${{ matrix.os }}
```
I have simplified the zip name here too.

shymega replied:
Yeah, something simpler would be best. I see you've done that now, so I'll close this convo.

shymega commented on .github/workflows/builds.yml:
```diff
@@ -1,40 +1,47 @@
       matrix:
         os: ["ubuntu:18.04", "ubuntu:20.04", "ubuntu:22.04"]
-    # Steps represent a sequence of tasks that will be executed as part of the job
+        cc: ["gcc"]
+        wayland: [false]
+        include:
+          - cc: "gcc"
+            cxx: "g++"
+          - os: "ubuntu:22.04"
+            cc: "clang"
+            cxx: "clang++"
+          - os: "ubuntu:22.04"
```
OK, thanks for clarifying.

shymega commented:
> Two more things as an afterthought: I was thinking of using the release functionality of this workflow to automagically trigger new releases on the Snap, Chocolately, Scoop, and Flatpak - these are registered, but not uploaded yet. Could we achieve this with a workflow trigger, or should I setup a AWS Lambda?

@Cimbali I'm curious about this. I can make a Lambda easily, and that should run fine. But I wasn't sure if we could achieve the same thing in GitHub natively.

Cimbali commented:
> > Two more things as an afterthought: I was thinking of using the release functionality of this workflow to automagically trigger new releases on the Snap, Chocolately, Scoop, and Flatpak - these are registered, but not uploaded yet. Could we achieve this with a workflow trigger, or should I setup a AWS Lambda?
> 
> @Cimbali I'm curious about this. I can make a Lambda easily, and that should run fine. But I wasn't sure if we could achieve the same thing in GitHub natively.

On github there is a wide catalog of actions, maybe some could facilitate these tasks though I haven’t looked into it too much. We could also roll our own but imho for things that do not run very frequently it’s a bit fiddly to debug. Definitely doable though. Otoh I don’t know anything about lambdas.

shymega commented:
> > > Two more things as an afterthought: I was thinking of using the release functionality of this workflow to automagically trigger new releases on the Snap, Chocolately, Scoop, and Flatpak - these are registered, but not uploaded yet. Could we achieve this with a workflow trigger, or should I setup a AWS Lambda?
> > 
> > 
> > @Cimbali I'm curious about this. I can make a Lambda easily, and that should run fine. But I wasn't sure if we could achieve the same thing in GitHub natively.
> 
> On github there is a wide catalog of actions, maybe some could facilitate these tasks though I haven’t looked into it too much. We could also roll our own but imho for things that do not run very frequently it’s a bit fiddly to debug. Definitely doable though. Otoh I don’t know anything about lambdas.

Yeah, Actions could work. Or a webhook. There are also 'Deployments'... but we don't release frequently (although my opinion is that we should release more frequently, with less 'bulk' of changes between releases), so as you say, it might be fiddly.

I use Lambdas for a bootloader project Telegram - not quite the same, but it parses GitHub events.

shymega commented:
@Cimbali So, seeing as we haven't heard from Povilas, but they _did_ approve the PR, I think we should merge.

The only thing I need is for the commits to be squashed, and then rebased against mainline.

Then I can merge, and we'll have a working CI!

shymega reviewed (approved):
It looks great! Ready for a merge I think.

whot commented:
> The only thing I need is for the commits to be squashed, and then rebased against mainline.

fwiw, if you select `squash and merge` gitlab will do that for you (though frustratingly it won't run the CI until *after* the merge)

shymega commented:
> > The only thing I need is for the commits to be squashed, and then rebased against mainline.
> 
> fwiw, if you select `squash and merge` gitlab will do that for you (though frustratingly it won't run the CI until _after_ the merge)

I would rather do it manually - it also means I can rebase against `master`. Also, did you mean GitLab or GitHub?

I've rebased locally now, just need to push to @Cimbali's fork - is that alright with you, @Cimbali?

Cimbali commented:
Yeah go for it.

whot commented:
> Also, did you mean GitLab or GitHub?

oops. same thing, they both have that button - though git*lab* allows you to rebase first (thus running the CI) and then merge+squash.

shymega commented:
> > Also, did you mean GitLab or GitHub?
> 
> oops. same thing, they both have that button - though git_lab_ allows you to rebase first (thus running the CI) and then merge+squash.

Yeah, that does make sense. Would be nice if GitHub had that feature!

shymega commented:
So, I've noted that when I merge this commit into my `rust` branch, there's a few merge conflicts. Yet, GitHub doesn't show these.

shymega commented:
Oh, d'y know. I realised I edited the workflow myself in that branch. Ignore me!

shymega commented:
@p12tic Can you resolve this merge conflict before we merge? Looks like your latest PRs broke some things in this PR.

We really need to merge this PR. Confusion is rising, and users who want Windows and macOS builds are being neglected.

Cimbali commented:
I’ve rebased your squashed commit on master @shymega, to fix the conflicts from ae3bfb2d

shymega commented:
> I’ve rebased your squashed commit on master @shymega, to fix the conflicts from [ae3bfb2](https://github.com/input-leap/input-leap/commit/ae3bfb2de8e1de2e1988b032322cf923c6cc1763)

Thanks, I'm going to merge this PR now tbh. It's been stalled for too long, and the maintainers have approved it.
